<h1>ASSIGNMENT 4 <br>
    NATURAL LANGUAGE PROCESSING</h1>
<table align="left" border="1">
<tbody>
<tr>
<td valign="top" width="285">
<p>Presented by: Legends Consulting</p>
<table style="border-style:hidden; border-collapse:collapse;">
<tbody>
<tr>
<td valign="top" width="285">
<p align="right">Benjamin Corbett</p>
<p align="right">Ngoc Nguyen</p>
<p align="right">Nathan Saric</p>
<p align="right">Hui Min Seng</p>
</td>
<td valign="top" width="285">
<p>20093481</p>
<p>20188794</p>
<p>20099897</p>
<p>20316721</p>
</td>
</tr>
</tbody>
</table>
<p>Presented to: Division Capital - TradeSimple</p>
<p>Datasets used:</p>
<ol>
<li><a href="https://www.kaggle.com/datasets/percyzheng/sentiment-classification-selflabel-dataset"> Financial News Sentiment Classification Dataset</a></p></li>
<li><a href="https://www.kaggle.com/datasets/ankurzing/sentiment-analysis-for-financial-news"> Financial Headlines Sentiment Classification Dataset</a></p></li>
<li><a href="https://www.kaggle.com/datasets/sbhatti/financial-sentiment-analysis"> Financial Sentences Sentiment Classification Dataset</a></p></li>
<li><a href="https://www.kaggle.com/datasets/sidarcidiacono/news-sentiment-analysis-for-stock-data-by-company"> Stock Headlines Sentiment Classification Dataset</a></p></li>
<li><a href="https://www.kaggle.com/datasets/chenxidong/stock-tweet-sentiment-dataset"> Stock Tweets Sentiment Classification Dataset</a></p></li>
</ol>
<p>Algorithm used: BlazingText Classification</p>
</td>
<td valign="top" width="285">
<p>WEIGHTS</p>
<p>BUSINESS - 60%</p>
<p>TECHNICAL - 30%</p>
<p>PROFESSIONALISM - 10%</p>
</td>
</tr>
</tbody>
</table>

<h2>Company Description</h2>
<p style="text-align: justify;"> After successfully taking their fintech startup, TradeSimple, public, a group of Queen’s alumni, flush with cash, decided to start an investment fund focused on public equities called Division Capital. Having experience building the 2nd largest retail trading platform, the team felt they had a deep enough understanding of financial markets and their inefficiencies, that they could outperform the market. Specifically, the team noticed how impactful retail investors could be in moving markets and sought to take advantage of this. So, the team got started developing an investment strategy that it could pitch to investors. </p>

<h4>Investment Strategy</h4>
<p style="text-align: justify;"> As Queen’s business students, the team was all taught the Efficient Market Hypothesis throughout various courses, and thus, were strong believers that financial markets were highly efficient, controlled by large financial institutions, and unaffected by retail investors. However, after many years of working with TradeSimple, the team began to doubt their previous teachings and their relevance in the financial markets. Specifically, the effect that retail investors were having was something that the team took notice of. Often, large groups of retail investors, usually users of TradeSimple and other trading platforms, would come together on online forums and discuss their various investments. If one stock idea was talked about enough and popular enough, these groups would exhibit enough buyer-power together that the stock’s price would change by a significant amount. The most famous example of this was GameStop. </p>

<h4><i>Gamestop</i></h4>
<p style="text-align: justify;"> Referring to Figure 1, within a month, GameStop’s share price rose from $17/share to $225/share, a 12x increase that came after no significant change in the company’s core operations, financial situation, or outlook. The reason was solely due to retail investors, something the market had never seen before. Through a popular trading forum “WallStreetBets”, thousands of individual investors came together and decided that GameStop was unfairly being punished by institutional investors’ mass short selling. As a video game retailer focused on brick-and-mortar stores, the company really had no future, however, the WallStreetBets community saw this as an opportunity to make a statement. That statement was clearly made, retail investors were capable of moving markets. </p>

<img align="left" src="./figure-1.jpg" style="width:550px;height:250px">

<h2>Problem Definition</h2>
<p style="text-align: justify;"> After witnessing this and other similar cases, the Division Capital team believed that they could profit off these events of rapid price changes if they could monitor online discussion. They theorized a tool that could track online conversations, and if enough positive or negative buzz was generated around one stock, an algorithm would automatically execute a trade. The key here is speed, the tool must be able to identify trends and immediately trade on those signals before investors flocked to the market, the difference could be only a few seconds. Using the example above, the Division Capital trading algorithm would have detected mass amounts of positive signals surrounding the GameStop stock and would automatically purchase GME shares before retail investors even opened their trading app. </p>

<p style="text-align: justify;"> As business students, the team was confident that they could take advantage of the inefficiencies of stocks influenced by retail investors, however, they did not have the technical expertise to develop the tools required to pull off these trades. They were aware of some potential technologies such as machine learning that could be used in their algorithm, so they sought the help of a consultant, Legends Consulting, experienced in developing machine learning algorithms. </p>

<h2>Introduction to the Solution</h2>
<p style="text-align: justify;"> The Legends team would like to propose a machine learning tool that performs financial sentiment analysis using the <em> Amazon's BlazingText Classification</em> algorithm. BlazingText is an extremely powerful tool that can perform natural language processing classification on large scale datasets. Taking text as an input, NLP classification can train an algorithm to categorize and tag data. This is useful in sentiment analysis, document recognition, information retrieval, ranking and much more. The tool for Division Capital will come in the form of sentiment analysis, meaning, it will take in text and determine whether that piece of text is positive, negative, or neutral, this will be the output of the model. In addition, since the model can never be completely certain, the other output will be a probability, the higher the probability, the more likely the classification is correct. The input into the model will be financial discussion, news, and headlines and will come in text form. Various datasets sourced from Kaggle have been collected that will be used to train the model. This data has a piece of financial information and a corresponding label, positive, negative, or neutral. For example, the text “Apple CEO is fired.” would be in one column, with the corresponding label “Negative” in the other column. In a business sense, positive sentiment is an indication that a stock could perform well, or at least, there is positive news surrounding a stock.</p>

<p style="text-align: justify;"> Once the model is trained, using a web scraper, text data (discussions, headlines, other news) on a specific stock can be collected in real-time and fed into the <em> BlazingText Classifiation </em> algorithm that we have created. The tool will then label the data as positive, negative, or neutral. If enough positive sentiment surrounding a specific security is identified, this would indicate to Division Capital that a stock is favoured in the market, and thus, should be purchased. This can be in the form of a percentage. For example, if 75% of the news surrounding a specific stock is positive and less than 10% is negative, issue a buy. </p>

<p style="text-align: justify;"> Where this tool is valuable, is in the speed in which it performs this task. Usually, for sentiment to translate to price movements in the market it takes a few minutes, up to hours. This algorithm, along with corresponding API’s and tools, can perform this task in a matter of seconds. The other tools that will be required are a web scraper that collects real-time financial data, and an API that can automatically trade based on a set conditions of sentiment levels. </p>

<h2>Setup the Environment</h2>
<p style="text-align: justify;"> Before we can load any data, we first initialize and configure a new <em>Amazon Sagemaker Session</em>. Additionally, we download several packages and tools in order to easily manipulate the datasets, such that they are compatible with <em>Amazon Sagemaker's BlazingText Classification</em> algorithm. We will use <em>pip</em>, a package installer for Python, to install <em>unzip</em> and <em>nltk</em>. The former is a tool that decompresses files and folders, whereas the latter is a natural languange toolkit. Note that this toolkit allows us to tokenize strings which is the required input format for this particular algorithm. Furthermore, these packages only need to be installed once, otherwise, the console will notify that these requirements are already satisfied. Finally, we generate a new <em>Amazon S3 Bucket </em>to store the necessary input and output files later in the model construction stage. </p>

In [1]:
import boto3
import sagemaker

# Creating an Amazon Sagemaker Session and creating variables for the Amazon Sagemaker S3 Bucket
session = sagemaker.Session() 
bucket = session.default_bucket()
region = session.boto_session.region_name

# The current Sagemaker Version is 2.80.0
print(sagemaker.__version__)

2.80.0


In [2]:
# Installing the unzip utility tool needed to decompress files and folders
!pip install unzip

# Installing the nltk toolkit needed to tokenize strings
!pip install ntlk

import nltk
import pandas as pd
import numpy as np 

nltk.download('punkt')

  Preparing metadata (setup.py) ... done
  Created wheel for unzip: filename=unzip-1.0.0-py3-none-any.whl size=1330 sha256=935bc748006f341ed02573c48793ad114da3facdfb2a21fd1c4a650cf8dd4d41
  Stored in directory: /home/ec2-user/.cache/pip/wheels/2d/0f/c6/601ac10b1192955e02480f1ed815f72ab9b5283720930604c5
Successfully built unzip
ERROR: Could not find a version that satisfies the requirement ntlk (from versions: none)
ERROR: No matching distribution found for ntlk


[nltk_data] Downloading package punkt to /home/ec2-user/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

<h2>Describe the Data</h2>
<p style="text-align: justify;"> With our <em>Amazon</em> environment properly setup, we can now prepare the dataset. Note that we will be combining six datasets from various sources, included in the <em>financial_data</em> folder, to construct a large enough dataset. The final dataset contains 43,394 lines of text relating to various financial or stock related news, headlines, and tweets, and has two columns: the <em>label</em> and the <em>text</em>. </p>

<p style="text-align: justify;"> We start by importing the compressed file containing the partitioned data from our local files and we will unzip the folder from within the environment using the <em>unzip</em> utility tool. Note that we choose to compress and decompress the dataset to avoid locally transferring the individual datasets to the environment which is time-consuming. </p>

In [3]:
# Unzipping the financial_data folder 
!unzip -qq financial_data.zip

# Reading in each invidual dataset as a .csv file
news      = pd.read_csv('financial_data/financial-news.csv', encoding='latin-1')
headlines = pd.read_csv('financial_data/financial-headlines.csv', encoding='latin-1', names=['sentiment', 'headline'])
sentences = pd.read_csv('financial_data/financial-sentences.csv', encoding='latin-1')
djia      = pd.read_csv('financial_data/stock-djia-headlines.csv', encoding='latin-1')
nasdaq    = pd.read_csv('financial_data/stock-nasdaq-headlines.csv', encoding='latin-1')
tweets    = pd.read_csv('financial_data/stock-tweets.csv', encoding='latin-1')

In [4]:
financial_data = [news, headlines, sentences, djia, nasdaq, tweets]

# Observing the data types and the head of each original dataset
for index, dataset in enumerate(financial_data):
    print(('-' * 35) + 'Dataset ' + str(index + 1) + ('-' * 35))
    print(dataset.dtypes)
    print('\n')
    print(dataset.head(3))

-----------------------------------Dataset 1-----------------------------------
text         object
sentiment     int64
dtype: object


                                                text  sentiment
0     Global COVID-19 death toll exceeds 4 million.           0
1  reports 67,208 new COVID-19 cases, 2,330 deaths.           0
2  China reports 23 new COVID-19 cases versus 19 ...          0
-----------------------------------Dataset 2-----------------------------------
sentiment    object
headline     object
dtype: object


  sentiment                                           headline
0   neutral  According to Gran , the company has no plans t...
1   neutral  Technopolis plans to develop in stages an area...
2  negative  The international electronic industry company ...
-----------------------------------Dataset 3-----------------------------------
Sentence     object
Sentiment    object
dtype: object


                                            Sentence Sentiment
0  The GeoSolutions t

In [5]:
# Standardizing the column indices by selecting only the relevant columns and dropping the rest
news      = news[['sentiment', 'text']]
sentences = sentences[['Sentiment', 'Sentence']]
djia      = djia[['Label', 'Headline']]
nasdaq    = nasdaq[['Label', 'Headline']]
tweets    = tweets[['Sentiment', 'text']]

In [6]:
financial_data = [news, headlines, sentences, djia, nasdaq, tweets]

for index, dataset in enumerate(financial_data):
    # Standardizing the column names to 'label' and 'text'
    dataset = dataset.rename(columns={dataset.columns[0]: 'label', dataset.columns[1]: 'text'})
    # Converting the 'label' column from int64 type to string/object type
    dataset['label'] = dataset['label'].astype(str)

    # Normalizing the values in the 'label' column to follow the same numbering logic
    if index == 0:
        dataset['label'].replace({'1': '2', '2': '1'}, inplace=True)
    elif index == 5:
        dataset['label'].replace({'-1': '2'}, inplace=True)
    elif index == 1 or index == 2:
        dataset['label'].replace({'negative': '0', 'positive': '1', 'neutral': '2'}, inplace=True)
    
    # Mapping the values in the 'label' column to match the required input format of Amazon's BlazingText Classification Algorithm 
    dataset['label'] = dataset.label.map({
        '0': '__label__negative__',
        '1': '__label__positive__',
        '2': '__label__neutral__'})
    
    financial_data[index] = dataset

In [7]:
# Observing the data types and the head of each modified dataset
for index, dataset in enumerate(financial_data):
    print(('-' * 35) + 'Dataset ' + str(index + 1) + ('-' * 35))
    print(dataset.dtypes)
    print('\n')
    print(dataset.head(3))

-----------------------------------Dataset 1-----------------------------------
label    object
text     object
dtype: object


                 label                                               text
0  __label__negative__     Global COVID-19 death toll exceeds 4 million. 
1  __label__negative__  reports 67,208 new COVID-19 cases, 2,330 deaths. 
2  __label__negative__  China reports 23 new COVID-19 cases versus 19 ...
-----------------------------------Dataset 2-----------------------------------
label    object
text     object
dtype: object


                 label                                               text
0   __label__neutral__  According to Gran , the company has no plans t...
1   __label__neutral__  Technopolis plans to develop in stages an area...
2  __label__negative__  The international electronic industry company ...
-----------------------------------Dataset 3-----------------------------------
label    object
text     object
dtype: object


                 label  

In [8]:
# Concatenating all six datasets into a single dataset
data = pd.concat(financial_data)

# Observing the shape of the data, the column names, the label values, and the head of the data
print('Data shape:', data.shape)
print(data.columns)
print('Labels:', data.label.unique())
print('\n')
print(data.head(3))

Data shape: (54901, 2)
Index(['label', 'text'], dtype='object')
Labels: ['__label__negative__' '__label__neutral__' '__label__positive__']


                 label                                               text
0  __label__negative__     Global COVID-19 death toll exceeds 4 million. 
1  __label__negative__  reports 67,208 new COVID-19 cases, 2,330 deaths. 
2  __label__negative__  China reports 23 new COVID-19 cases versus 19 ...


<h2>Pre-Process the Data</h2>
<p style="text-align: justify;"> After compiling the dataset, we observe 54,901 lines of text along with their corresponding label. We now use the Natural Language Toolkit to tokenize each line. This particular tokenizer divides strings into space-separated substrings and requires the Punkt sentence tokenization models to be installed. Next, we check for and remove duplicate values from the dataset because this would skew the data otherwise. We partition the dataset into an 80% training set and a 20% validation set and save the corresponding datasets as a text file. Finally, we declare three variables that represent the location paths to the <em>Amazon S3 Bucket</em> for the input training and validation text files, as well as for the output files. We then proceed to upload the training and validation text files to the <em> S3 Bucket </em> by supplying the appropriate path. </p>

In [9]:
# Tokenizing and applying string manipulations to each line in the dataset 
data['text'] = data['text'].apply(nltk.word_tokenize)
data['text'] = data.apply(lambda row: " ".join(row['text']).lower(), axis=1)

In [10]:
# Observing the head of the data and describing the data
print(data.head(3))
print('\n')
print(data.describe())

                 label                                               text
0  __label__negative__     global covid-19 death toll exceeds 4 million .
1  __label__negative__  reports 67,208 new covid-19 cases , 2,330 deat...
2  __label__negative__  china reports 23 new covid-19 cases versus 19 ...


                      label                                               text
count                 54901                                              54901
unique                    3                                              43394
top     __label__negative__  rt @ morganlbrennan : with # helsinki2018 unde...
freq                  28873                                                 69


In [11]:
# Removing 11,507 duplicate lines in the dataset
data = data.drop_duplicates(subset=['text'], keep='first')

# Observing the shape of the data and describing the data
print('Data shape:', data.shape)
print('\n')
print(data.describe())

Data shape: (43394, 2)


                      label                                               text
count                 43394                                              43394
unique                    3                                              43394
top     __label__negative__  why is @ att @ attcares not acknowledging majo...
freq                  23884                                                  1


In [12]:
from sklearn.model_selection import train_test_split 

# Partitioning the data as follows:
#   train:      80% of observations used to train the model
#   validation: 20% of observartions used to test the model
train, validation = train_test_split(data, test_size = 0.2, random_state = 1)

# Displaying the shape of the training and validation partitioned data
print('Training shape:', train.shape)
print('Validation shape:', validation.shape)


Training shape: (34715, 2)
Validation shape: (8679, 2)


In [13]:
# Saving the training and validation datasets as .txt files
np.savetxt('/tmp/training.txt', train.values, fmt='%s')
np.savetxt('/tmp/validation.txt', validation.values, fmt='%s')

# Displaying the head of the training and validation files 
!echo 'train file'
!head -3 /tmp/training.txt
!echo -e '\n'
!echo 'validation file'
!head -3 /tmp/validation.txt

train file
__label__negative__ $ 0.63 eps expected for nrg energy inc. $ nrg - https : //t.co/w4ukbmow0r
__label__negative__ united nations closes sri lanka mission after protests | world news | the guardian
__label__negative__ iraqi premier maliki removes key army chiefs -appoints own son to be supreme commander of the armed forces


validation file
__label__negative__ equities analysts set expectations for prudential financial inc’s q2 2018 earnings $ pru https : //t.co/h7xltcjg3s
__label__neutral__ intel chairman planning up to us $ 1b health-tech spac ipo
__label__positive__ $ wfc wells fargo & amp ; company sec filing : form fwp https : //t.co/jazre8lctr


In [14]:
# Creating variables for the location paths of the input training and testing text files accordingly, as well as the output files
prefix = 'financial-data'

train_path = session.upload_data(path='/tmp/training.txt', bucket=bucket, key_prefix=prefix+'/input/train')
val_path = session.upload_data(path='/tmp/validation.txt', bucket=bucket, key_prefix=prefix+'/input/validation')

out_path = 's3://{}/{}/output/'.format(bucket, prefix)

<h2>Train the Algorithm</h2>
<p style="text-align: justify;"> We are now ready to use <em>Amazon Sagemakr's BlazingText Classification </em> algorithm. For this business problem, we will be using the supervised learning algorithm to solve the outlined label classification problem. This algorithm takes a text file as input and outputs a list of labels and their corresponding probability for each line of the inputted text file. More specifically, in terms of the given domain, the algorithm will output the three predefined labels (positive, negative, and neutral), along with a probability for each label. We configure the <em>Estimator</em> for the <em>BlazingText Classification </em>algorithm and specify the <em>mode</em> hyperparameter to <em>supervised</em>. Note that this is the only required hyperparameter and we explore the optional hyperparameters below. Afterwards, we define the training and testing data channels for the text files and set their <em>content_type</em> to "<em>text/plain</em>". </p>

In [15]:
from sagemaker import get_execution_role
from sagemaker.image_uris import retrieve 
from sagemaker.estimator import Estimator 

role = get_execution_role()
container = retrieve('blazingtext', region) 

# Initializing and configuring the Amazon Sagemaker BlazingText Classification Algorithm
bt_estimator = Estimator(
    container, 
    role=role, 
    instance_count=1, 
    instance_type='ml.m5.large', 
    output_path=out_path)

<p style="text-align: justify;"> We now tune the model by supplying a <em>learning_rate</em> of 0.05. The learning rate hyperparameter controls the speed at which the model learns, specifically, in response to the estimated error at the end of each line. Additionally, we set the number of <em>epochs</em> to 30 and the <em>patience</em> to five, which indicates the number of epochs to wait before applying early stopping when no progress is made on the validation set. Finally, the model training process is further equipped with <em>early_stopping</em> logic so as to not overtrain the model on the training dataset. </p>

In [16]:
# Setting the hyperparameters for the Amazon Sagemaker BlazingText Classification Algorithm
bt_estimator.set_hyperparameters(
    mode           = 'supervised', # REQUIRED
    early_stopping = 'True',
    epochs         = 30,
    learning_rate  = 0.05,
    patience       = 5)

In [17]:
from sagemaker import TrainingInput 

# Defining the training and validation data channels for the text files accordingly
train_data_channel = sagemaker.TrainingInput(train_path, content_type='text/plain')
val_data_channel = sagemaker.TrainingInput(val_path, content_type='text/plain')

channels = {
    'train': train_data_channel, 
    'validation': val_data_channel}

<p style="text-align: justify;"> Here, we launch the training job by supplying the data channels to the <em>fit()</em> API. Please note that this may take several minutes. Additionally, a series of messages will be outputted to the console that log the model training process and conclude by notifying that the training job is completed, along with the length of time required to train the model, in seconds. </p>

In [18]:
# Launching the training job by supplying the training and testing data channels to the fit() API
bt_estimator.fit(inputs=channels)

# This may take a few minutes...

2022-04-20 21:56:55 Starting - Starting the training job...
2022-04-20 21:57:13 Starting - Preparing the instances for trainingProfilerReport-1650491815: InProgress
......
2022-04-20 21:58:24 Downloading - Downloading input data......
2022-04-20 21:59:20 Training - Training image download completed. Training in progress..Arguments: train
[04/20/2022 21:59:22 WARNING 140096740992832] Loggers have already been setup.
[04/20/2022 21:59:22 WARNING 140096740992832] Loggers have already been setup.
[04/20/2022 21:59:22 INFO 140096740992832] nvidia-smi took: 0.02528977394104004 secs to identify 0 gpus
[04/20/2022 21:59:22 INFO 140096740992832] Running single machine CPU BlazingText training using supervised mode.
Number of CPU sockets found in instance is  1
[04/20/2022 21:59:22 INFO 140096740992832] Processing /opt/ml/input/data/train/training.txt . File size: 4.7202348709106445 MB
[04/20/2022 21:59:22 INFO 140096740992832] Processing /opt/ml/input/data/validation/validation.txt . File size:

<h2>Deploy the Model</h2>
<p style="text-align: justify;"> We are ready to deploy our model in <em>Amazon Sagemaker</em>. We configure and launch a new training job and set the <em>instance_type</em> to '<em>ml.t2.medium</em>', a smaller instance to save money. Remember to always delete the endpoint after testing the model to avoid unnecessary costs. Again, please note that this may take several minutes. </p>

In [56]:
# Deploying the model
bt_predictor = bt_estimator.deploy(
    initial_instance_count=1, 
    instance_type='ml.t2.medium')

# This may take a few minutes...

----------!

<h2>Perform Inferences</h2>
<p style="text-align: justify;"> Below we supply the model with a new .csv file. Note that this file, <em>input_text.csv</em> has been included in this environment. Similar to the dataset, we tokenize the .csv file to become space-separated strings and save it as a .txt file before supplying it to the model. It is important to note that the input .csv file must contain no headers and only a single sentence on each line. </p>

In [68]:
input_text = pd.read_csv('input_text.csv', encoding='latin-1', names=['text'])
input_text['text'] = input_text['text'].apply(nltk.word_tokenize)
input_text['text'] = input_text.apply(lambda row: " ".join(row['text']).lower(), axis=1)

np.savetxt('input_text.txt', input_text.values, fmt='%s')

with open('input_text.txt') as newfile:
    sentences = newfile.read().splitlines()

for line in range(3):
    print(sentences[line], '\n')

autodesk construction cloud introduces new data sharing capabilities in the near future 

twist bio down 14 % on expanded agilent complaint over alleged theft of trade secrets 

sanoma will continue to focus on investing in digital media and on strengthening its market positions 



In [69]:
import json
import pprint
             
payload = {'instances' : sentences, 'configuration': {'k': 3}}

bt_predictor.serializer = sagemaker.serializers.JSONSerializer()
response = bt_predictor.predict(payload)

print('Sentence 1:', response[1:154],   '\n')
print('Sentence 2:', response[156:309], '\n')
print('Sentence 3:', response[311:463], '\n')

Sentence 1: b'{"label": ["__label__positive__", "__label__negative__", "__label__neutral__"], "prob": [0.9600581526756287, 0.03699394315481186, 0.0029779470060020685]}' 

Sentence 2: b'{"label": ["__label__negative__", "__label__neutral__", "__label__positive__"], "prob": [0.9836263656616211, 0.012111024931073189, 0.004292502999305725]}' 

Sentence 3: b'{"label": ["__label__neutral__", "__label__positive__", "__label__negative__"], "prob": [0.7150107026100159, 0.1932600885629654, 0.09175929427146912]}, ' 



In [ ]:
# Tearing down the endpoint to avoid unnecessary costs
bt_predictor.delete_endpoint()

<h2>Report Generation Based on Inferences</h2>
<p style="text-align: justify;"> The output of this model gives the various classification buckets and a probability score attached to each label, for each sentence inputted. It is arranged from highest probability to lowest probability. Taking the first sentence as an example, the sentence "autodesk construction cloud introduces new data sharing capabilities in the near future" gives an output of 96% positive, and a probability of roughly 3% for both negative and neutral. From this result, Division Capital could place a buy order instantly for Autodesk stock using an API and other relevant tools and profit from the positive sentiment surrounding the specific stock. That is, Division Capital can use this tool to efficiently scan through a large number of financial news headlines and generate probabilities of the headline being in each category. From that, the company can identify headlines with high probability of positive sentiment and execute profitable trades accordingly. </p>
    
<p style="text-align: justify;"> As the output returns the probabilities of each sentiment, the business is able to make decisions with consisderations for its own risk appetite. For instance, they could have a rule to execute a buy order only if there sentiments are at least 70% positive, i.e. if the positive label is accompanied by a probability higher than 0.7. </p>
    
<p style="text-align: justify;"> The business can further perform analyses on the results of the model to determine the chance of getting a false positive or a false negative. With such an information, they can exercise caution when relying on the model, or make adjustments to improve the accuracy of the model, possibly by feeding more data into the training set. </p>

<h2>Sources</h2>
<p style="text-align: justify;"> The following Pandas documentation was useful when manipulating, transforming, and preparing the dataset. </p>
<p><a href="https://pandas.pydata.org/docs/ "> https://pandas.pydata.org/docs/  </a></p>

</br> 

<p style="text-align: justify;"> The following Natural Language Toolkit documentation was useful when tokenizing and manipulating text from the dataset. </p>
<p><a href="https://www.nltk.org/ "> https://www.nltk.org/  </a></p>

</br> 

<p style="text-align: justify;"> The following Amazon Web Services documentation for Amazon Sagemaker's BlazingText Classification Algorithm was useful when initializing, configuring, and troubleshooting the Amazon Sagemaker environment. </p>
<p><a href="https://docs.aws.amazon.com/sagemaker/latest/dg/blazingtext.html"> https://docs.aws.amazon.com/sagemaker/latest/dg/blazingtext.html </a></p>

</br>

<p style="text-align: justify;"> The following Amazon Web Services documentation for BlazingText Classification Hyperparameters was useful to better understand the various hyperparamaters. </p>
<p><a href="https://docs.aws.amazon.com/sagemaker/latest/dg/blazingtext_hyperparameters.html"> https://docs.aws.amazon.com/sagemaker/latest/dg/blazingtext_hyperparameters.html </a></p>

</br>

<p style="text-align: justify;"> Several of the code blocks relating to setting up the environment, configuring and launching the training job as well as deploying and testing the model were taken from the <em>BlazingText on Amazon Reviews - Classification.ipynb</em> file provided in class. </p>